packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -388,6 +410,28 @@ class SliverChildBuilderDelegate extends SliverChildDelegate {
   /// Defaults to providing an index for each widget.
   final SemanticIndexCallback semanticIndexCallback;
 
+  /// Called to find the new index of a child based on its key in case of reordering.
+  ///
+  /// If not provided, a child widget may not map to its existing [RenderObject]
+  /// when the order in which children are returned from [builder] changes.
+  /// This may result in state-loss.
+  final Function findChildIndex;
+
+  @override
+  int findIndexByKey(Key key) {
+    if (findChildIndex == null)
+      return -1;
+    assert(key != null);
```

Comment:
the rest of this is almost identical to another method below. can we avoid the code duplication somehow?

Reply:
I am not sure how we can do this, these two class are sibling, and the way this function unwrap the key is depends on builder function in these two child classes. It is probably better to leave this function implementation to the individual child class